File: tests/always-on/arms/lean/skills/search/scripts/jobs_md.py
#!/usr/bin/env python3
"""jobs.md — the pipeline as a readable markdown record (the one record).

Script-owned: regenerated in full on every save, organized by stage so the
file reads as the board. Humans read it anywhere; changes go through the
owning scripts (search_ats.py appends, update_job.py moves, record_verdict.py
judges) so the loud-fail duplicate key survives the storage change.

Design (settled at #14's gate): one labeled block per role — a 20-field row
is unreadable as a table row. SQLite is gone as persistent storage; in-run
set logic uses plain dicts. The escape hatch stays: at a few hundred roles,
parsing markdown to answer questions gets silly and the pipeline moves back
to a real table — reverse on evidence.
"""
import os, re
from datetime import datetime, timezone

STAGES = ["To Review", "Interested", "Applied", "Interviewing", "Offer"]
DISMISSED = "Dismissed"
NOTES = "Search notes"

# block field label -> row key (order here = write order)
FIELDS = [
    ("URL", "url"), ("Location", "location"), ("Posted", "posted_at"),
    ("Seen", "seen_at"), ("Updated", "updated_at"),
    ("Verdict", "fit_verdict"), ("Score", "fit_score"), ("Reason", "fit_reason"),
    ("Dealbreakers", "dealbreakers"), ("Track", "track"), ("Flags", "flags"),
    ("Sim", "jd_sim"), ("JD", "jd_file"), ("Company file", "company_file"),
    ("Evaluated", "evaluated_at"), ("Was", "was_stage"), ("Dismissed", "dismiss_note"),
]
LABEL_TO_KEY = {l: k for l, k in FIELDS}


def canon(s):
    s = re.sub(r"[^a-z0-9 ]", " ", (s or "").lower())
    s = re.sub(r"\b(inc|llc|corp|labs|technologies|company|the)\b", " ", s)
    return re.sub(r"\s+", " ", s).strip()


def key(row):
    return (canon(row["company"]), canon(row["title"]))


def now_iso():
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def path(workspace):
    return os.path.join(workspace, "jobs.md")


def load(workspace):
    """Parse jobs.md -> list of row dicts. Absent file -> empty pipeline.
    Unknown ## sections (e.g. Search notes) are NOT stages — their content
    never parses as roles."""
    p = path(workspace)
    rows, cur_stage, row = [], None, None
    if not os.path.exists(p):
        return rows
    for line in open(p, encoding="utf-8"):
        line = line.rstrip("\n")
        m = re.match(r"^##\s+(.+?)\s*$", line)
        if m and not line.startswith("###"):
            name = m.group(1).strip()
            cur_stage = name if name in STAGES or name == DISMISSED else None
            continue
        m = re.match(r"^###\s+(.+?)\s*$", line)
        if m and cur_stage:
            head = m.group(1)
            company, _, title = head.partition(" — ")
            row = {"company": company.strip(), "title": title.strip(),
                   "stage": cur_stage if cur_stage != DISMISSED else None,
                   "dismissed": cur_stage == DISMISSED}
            rows.append(row)
            continue
        m = re.match(r"^-\s+([^:]+):\s*(.*)$", line)
        if m and row is not None:
            k = LABEL_TO_KEY.get(m.group(1).strip())
            if k:
                row[k] = m.group(2).strip() or None
    for r in rows:
        if r["dismissed"]:
            r["stage"] = r.get("was_stage") or "To Review"
        r.setdefault("fit_score", None)
        if r.get("fit_score"):
            try:
                r["fit_score"] = int(r["fit_score"])
            except ValueError:
                r["fit_score"] = None
    return rows


def load_notes(workspace):
    """The free-text ## Search notes section (script-preserved, agent-written:
    market signals, lane findings — search OUTPUT, so it lives here and never
    in criteria.md, which is input)."""
    p = path(workspace)
    if not os.path.exists(p):
        return ""
    m = re.search(r"^## Search notes\s*$\n(.*)\Z", open(p, encoding="utf-8").read(),
                  re.M | re.S)
    return m.group(1).strip() if m else ""


def append_note(workspace, text):
    notes = load_notes(workspace)
    stamp = datetime.now(timezone.utc).date().isoformat()
    block = f"### {stamp}\n\n{text.strip()}"
    save(workspace, load(workspace), notes=(notes + "\n\n" + block).strip() if notes else block)


def save(workspace, rows, notes=None):
    """Regenerate jobs.md in full: stage sections in board order, dismissed
    last with reasons. A duplicate canonical key is a HARD error — the
    UNIQUE constraint, ported."""
    seen = {}
    for r in rows:
        k = key(r)
        if k in seen:
            raise SystemExit(f"jobs.md: duplicate role {r['company']} — {r['title']} "
                             f"(canonical collision with {seen[k]['company']} — {seen[k]['title']})")
        seen[k] = r
    out = ["# Pipeline", "",
           "*The record. Script-written (search sweeps, update_job.py moves,",
           "record_verdict.py judges) — read it anywhere; change it via chat so",
           "the duplicate-key check can protect it. Dismissed roles keep their",
           "history at the bottom; nothing is ever deleted.*", ""]
    active = [r for r in rows if not r.get("dismissed")]
    out.append(f"**Active: {len(active)}** · dismissed: {len(rows) - len(active)} · updated {now_iso()[:10]}")
    out.append("")
    for stage in STAGES:
        block = [r for r in active if r.get("stage") == stage]
        if not block:
            continue
        out.append(f"## {stage}")
        out.append("")
        for r in sorted(block, key=lambda x: (-(x.get("fit_score") or 0), x["company"].lower(), x["title"].lower())):
            out.extend(_block(r))
    gone = [r for r in rows if r.get("dismissed")]
    if gone:
        out.append(f"## {DISMISSED}")
        out.append("")
        for r in sorted(gone, key=lambda x: (x["company"].lower(), x["title"].lower())):
            r["was_stage"] = r.get("stage") or "To Review"
            out.extend(_block(r, dismissed=True))
    if notes is None:
        notes = load_notes(workspace)   # a plain save never destroys the notes
    if notes:
        out += ["## Search notes", "", notes, ""]
    with open(path(workspace), "w", encoding="utf-8") as f:
        f.write("\n".join(out).rstrip() + "\n")


def _block(r, dismissed=False):
    lines = [f"### {r['company']} — {r['title']}"]
    for label, k in FIELDS:
        if k == "was_stage" and not dismissed:
            continue
        if k == "dismiss_note":
            v = r.get("dismiss_note") or (r.get("dismiss_reason") if dismissed else None)
        else:
            v = r.get(k)
        if v not in (None, ""):
            lines.append(f"- {label}: {v}")
    lines.append("")
    return lines


def find(rows, company, title_frag):
    """Match on canonical company + title substring; exact-title wins over
    substring collisions (ported from update_job.py). Returns matches."""
    ck, tf = canon(company), canon(title_frag)
    hits = [r for r in rows if canon(r["company"]) == ck and tf in canon(r["title"])]
    if len(hits) > 1:
        exact = [r for r in hits if canon(r["title"]) == tf]
        if len(exact) == 1:
            return exact
    return hits

File: packages/checkers/src/jobs-md.mjs
// A faithful JS port of skills/search/scripts/jobs_md.py — jobs.md as the
// pipeline's one readable record. Runs in Node AND in the browser; no
// node:fs, no node:path. Callers inject an `io` object (see README.md "The
// io interface") and, for save()/append_note(), a `now` clock (defaults to
// `() => new Date()`) so the parity test can freeze time exactly the way
// the real script's `datetime.now(timezone.utc)` is frozen for comparison.
import { join } from "./path-util.mjs";
import { pyInt, codePointCompare, pyRstrip, pyStrip, restoreLineSeparators, PY_S } from "./py-text.mjs";

// The Python source's `\s` in these patterns is Python's own whitespace
// set, which INCLUDES U+2028/U+2029 — but by the time text reaches here,
// `universalNewlines()` has already swapped any real U+2028/U+2029 for
// PY_S's own sentinel code points (so they survive JS's `^`/`$` as
// ordinary characters, not LineTerminators — see py-text.mjs). A NATIVE
// JS `\s` in these regexes would therefore silently stop matching
// whitespace Python's `\s*` still would (the corpus's
// `r2-rv-nel-and-u2028-in-fields` case: a field value like
// "Location:  x" where the second "space" is really a U+2029 — Python's
// `\s*` consumes it as part of the label/value separator; a bare JS `\s*`
// would leave the sentinel glued to the front of the captured value
// instead). PY_S is JS's `\s` set adjusted to match Python's exactly,
// AND it explicitly includes the two sentinels for this reason.
const HEADING2_RE = new RegExp(`^##${PY_S}+(.+?)${PY_S}*$`);
const HEADING3_RE = new RegExp(`^###${PY_S}+(.+?)${PY_S}*$`);
const FIELD_RE = new RegExp(`^-${PY_S}+([^:]+):${PY_S}*(.*)$`);

export const STAGES = ["To Review", "Interested", "Applied", "Interviewing", "Offer"];
export const DISMISSED = "Dismissed";

const FIELDS = [
  ["URL", "url"], ["Location", "location"], ["Posted", "posted_at"],
  ["Seen", "seen_at"], ["Updated", "updated_at"],
  ["Verdict", "fit_verdict"], ["Score", "fit_score"], ["Reason", "fit_reason"],
  ["Dealbreakers", "dealbreakers"], ["Track", "track"], ["Flags", "flags"],
  ["Sim", "jd_sim"], ["JD", "jd_file"], ["Company file", "company_file"],
  ["Evaluated", "evaluated_at"], ["Was", "was_stage"], ["Dismissed", "dismiss_note"],
];
const LABEL_TO_KEY = new Map(FIELDS);

export function canon(s) {
  let out = (s || "").toLowerCase().replace(/[^a-z0-9 ]/g, " ");
  out = out.replace(/\b(inc|llc|corp|labs|technologies|company|the)\b/g, " ");
  return out.replace(/\s+/g, " ").trim();
}

export function key(row) {
  return `${canon(row.company)}\u0000${canon(row.title)}`;
}

// datetime.now(timezone.utc).isoformat(timespec="seconds") — e.g.
// "2026-09-23T12:34:56+00:00" (note the "+00:00", not "Z").
export function nowIso(now = () => new Date()) {
  const d = now();
  const pad = (n) => String(n).padStart(2, "0");
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}+00:00`
  );
}

export function path(workspace) {
  return join(workspace, "jobs.md");
}

export async function load(io, workspace) {
  const p = path(workspace);
  const rows = [];
  if (!(await io.exists(p))) return rows;
  const text = await io.readFile(p);
  const lines = text.split(/\r?\n/);
  let curStage = null;
  let row = null;
  for (const raw of lines) {
    const line = raw.replace(/\n$/, "");
    let m = line.match(HEADING2_RE);
    if (m && !line.startsWith("###")) {
      const name = pyStrip(m[1]);
      curStage = STAGES.includes(name) || name === DISMISSED ? name : null;
      continue;
    }
    m = line.match(HEADING3_RE);
    if (m && curStage) {
      const head = m[1];
      const sepIdx = head.indexOf(" — ");
      const company = sepIdx === -1 ? head : head.slice(0, sepIdx);
      const title = sepIdx === -1 ? "" : head.slice(sepIdx + 3);
      row = {
        company: pyStrip(company),
        title: pyStrip(title),
        stage: curStage !== DISMISSED ? curStage : null,
        dismissed: curStage === DISMISSED,
      };
      rows.push(row);
      continue;
    }
    m = line.match(FIELD_RE);
    if (m && row !== null) {
      const k = LABEL_TO_KEY.get(pyStrip(m[1]));
      if (k) row[k] = pyStrip(m[2]) || null;
    }
  }
  for (const r of rows) {
    if (r.dismissed) r.stage = r.was_stage || "To Review";
    if (!("fit_score" in r)) r.fit_score = null;
    if (r.fit_score) {
      // Python: `int(r["fit_score"])`, ValueError caught -> None. This is
      // a DIFFERENT int conversion than argparse's type=int (argx.mjs) —
      // it's a plain `int(str)` with no error surfaced to the caller.
      r.fit_score = pyInt(r.fit_score);
    }
  }
  return rows;
}

export async function loadNotes(io, workspace) {
  const p = path(workspace);
  if (!(await io.exists(p))) return "";
  const text = await io.readFile(p);
  const m = text.match(new RegExp(`^## Search notes${PY_S}*$\\n([\\s\\S]*)$`, "m"));
  return m ? pyStrip(m[1]) : "";
}

export async function appendNote(io, workspace, text, now = () => new Date()) {
  const notes = await loadNotes(io, workspace);
  const stamp = now().toISOString().slice(0, 10);
  const block = `### ${stamp}\n\n${pyStrip(text)}`;
  const rows = await load(io, workspace);
  await save(io, workspace, rows, { notes: notes ? pyStrip(`${notes}\n\n${block}`) : block, now });
}

export class DuplicateKeyError extends Error {}

export async function save(io, workspace, rows, { notes = undefined, now = () => new Date() } = {}) {
  const seen = new Map();
  for (const r of rows) {
    const k = key(r);
    if (seen.has(k)) {
      const other = seen.get(k);
      throw new DuplicateKeyError(
        `jobs.md: duplicate role ${r.company} — ${r.title} (canonical collision with ${other.company} — ${other.title})`
      );
    }
    seen.set(k, r);
  }
  const out = [
    "# Pipeline",
    "",
    "*The record. Script-written (search sweeps, update_job.py moves,",
    "record_verdict.py judges) — read it anywhere; change it via chat so",
    "the duplicate-key check can protect it. Dismissed roles keep their",
    "history at the bottom; nothing is ever deleted.*",
    "",
  ];
  const active = rows.filter((r) => !r.dismissed);
  const iso = nowIso(now);
  out.push(`**Active: ${active.length}** · dismissed: ${rows.length - active.length} · updated ${iso.slice(0, 10)}`);
  out.push("");
  for (const stage of STAGES) {
    const block = active.filter((r) => r.stage === stage);
    if (!block.length) continue;
    out.push(`## ${stage}`);
    out.push("");
    const sorted = block.slice().sort((a, b) => {
      const sa = -(a.fit_score || 0);
      const sb = -(b.fit_score || 0);
      if (sa !== sb) return sa - sb;
      const ca = a.company.toLowerCase();
      const cb = b.company.toLowerCase();
      if (ca !== cb) return codePointCompare(ca, cb);
      return codePointCompare(a.title.toLowerCase(), b.title.toLowerCase());
    });
    for (const r of sorted) out.push(..._block(r, false));
  }
  const gone = rows.filter((r) => r.dismissed);
  if (gone.length) {
    out.push(`## ${DISMISSED}`);
    out.push("");
    const sorted = gone.slice().sort((a, b) => {
      const ca = a.company.toLowerCase();
      const cb = b.company.toLowerCase();
      if (ca !== cb) return codePointCompare(ca, cb);
      return codePointCompare(a.title.toLowerCase(), b.title.toLowerCase());
    });
    for (const r of sorted) {
      r.was_stage = r.stage || "To Review";
      out.push(..._block(r, true));
    }
  }
  let finalNotes = notes;
  if (finalNotes === undefined) finalNotes = await loadNotes(io, workspace);
  if (finalNotes) {
    out.push("## Search notes", "", finalNotes, "");
  }
  const text = restoreLineSeparators(pyRstrip(out.join("\n"))) + "\n";
  await io.writeFile(path(workspace), text);
}

function _block(r, dismissed) {
  const lines = [`### ${r.company} — ${r.title}`];
  for (const [label, k] of FIELDS) {
    if (k === "was_stage" && !dismissed) continue;
    let v;
    if (k === "dismiss_note") {
      v = r.dismiss_note || (dismissed ? r.dismiss_reason : undefined);
    } else {
      v = r[k];
    }
    if (v !== undefined && v !== null && v !== "") lines.push(`- ${label}: ${v}`);
  }
  lines.push("");
  return lines;
}

export function find(rows, company, titleFrag) {
  const ck = canon(company);
  const tf = canon(titleFrag);
  const hits = rows.filter((r) => canon(r.company) === ck && canon(r.title).includes(tf));
  if (hits.length > 1) {
    const exact = hits.filter((r) => canon(r.title) === tf);
    if (exact.length === 1) return exact;
  }
  return hits;
}
